src/msg/async/ProtocolV2.cc

```diff
@@ -234,93 +720,202 @@ void ProtocolV2::send_message(Message *m) {
+
   switch (state) {
     case START_CONNECT:
-      CONTINUATION_RUN(CONTINUATION(send_client_banner));
+      run_continuation(CONTINUATION(start_client_banner_exchange));
       break;
     case START_ACCEPT:
-      CONTINUATION_RUN(CONTINUATION(send_server_banner));
+      run_continuation(CONTINUATION(start_server_banner_exchange));
       break;
-    case OPENED:
-      CONTINUATION_RUN(CONTINUATION(wait_message));
+    case READY:
+      run_continuation(CONTINUATION(read_frame));
       break;
     case THROTTLE_MESSAGE:
-      CONTINUATION_RUN(CONTINUATION(throttle_message));
+      run_continuation(CONTINUATION(throttle_message));
       break;
     case THROTTLE_BYTES:
-      CONTINUATION_RUN(CONTINUATION(throttle_bytes));
+      run_continuation(CONTINUATION(throttle_bytes));
       break;
     case THROTTLE_DISPATCH_QUEUE:
-      CONTINUATION_RUN(CONTINUATION(throttle_dispatch_queue));
+      run_continuation(CONTINUATION(throttle_dispatch_queue));
       break;
     default:
       break;
   }
 }
 
+Message *ProtocolV2::_get_next_outgoing(bufferlist *bl) {
+  Message *m = 0;
+  if (!out_queue.empty()) {
+    map<int, list<pair<bufferlist, Message *>>>::reverse_iterator it =
+        out_queue.rbegin();
+    ceph_assert(!it->second.empty());
+    list<pair<bufferlist, Message *>>::iterator p = it->second.begin();
+    m = p->second;
+    if (bl) {
+      bl->swap(p->first);
+    }
+    it->second.erase(p);
+    if (it->second.empty()) {
+      out_queue.erase(it->first);
+    }
+  }
+  return m;
+}
+
+ssize_t ProtocolV2::write_message(Message *m, bufferlist &bl, bool more) {
+  FUNCTRACE(cct);
+  ceph_assert(connection->center->in_thread());
+  m->set_seq(++out_seq);
+
+  connection->lock.lock();
+  uint64_t ack_seq = in_seq;
+  ack_left = 0;
+  connection->lock.unlock();
+
+  ceph_msg_header &header = m->get_header();
+  ceph_msg_footer &footer = m->get_footer();
+
+  ceph_msg_header2 header2{header.seq,        header.tid,
+                           header.type,       header.priority,
+                           header.version,    header.front_len,
+                           header.middle_len, 0,
+                           header.data_len,   header.data_off,
+                           ack_seq,           footer.front_crc,
+                           footer.middle_crc, footer.data_crc,
+                           footer.flags,      header.compat_version,
+                           header.reserved,   0};
+
+  if (messenger->crcflags & MSG_CRC_HEADER) {
+    header2.header_crc =
+        ceph_crc32c(0, (unsigned char *)&header2,
+                    sizeof(header2) - sizeof(header2.header_crc));
+  }
+
+  bufferlist flat_bl;
+  if ((bl.length() <= ASYNC_COALESCE_THRESHOLD) && (bl.buffers().size() > 1)) {
+    for (const auto &pb : bl.buffers()) {
+      flat_bl.append((char *)pb.c_str(), pb.length());
+    }
+  } else {
+    flat_bl.claim_append(bl);
+  }
+
+  sign_payload(flat_bl);
+  encrypt_payload(flat_bl);
```

Comment:
Do we have any business in providing *confidentiality* **without** *authentication & integrity*? Separating into two methods suggests they are somewhat independent.

I've asked because taking a look on [AES-CMAC](https://tools.ietf.org/html/rfc4493).